### PR TITLE
problem: concurrent rw on state db stateobjects map

### LINKED ISSUE #284: Geth 3.5.0 Constantly Crashing - concurrent map read and map write
Our Geth 3.5.0 instance seems to be crashing after just a few hours.

Running on Debain, 32 GB RAM.

go version go1.7.3 linux/amd64

I0629 15:19:04.117996 eth/handler.go:295] Peer f3675296843aec4c [eth/63]: timed out fork-check, dropping
I0629 15:19:29.425872 eth/handler.go:295] Peer e3e0cf816cfea25e [eth/63]: timed out fork-check, dropping
I0629 15:19:49.120562 core/blockchain.go:947] imported 1 block(s) (0 queued 0 ignored) including 1 txs in 5.006234ms. #4001720 [b1d7ceb2 / b1d7ceb2]
fatal error: concurrent map read and map write

goroutine 46110 [running]:
runtime.throw(0xf13fb3, 0x21)
	/home/travis/.gimme/versions/go1.8.linux.amd64/src/runtime/panic.go:596 +0x95 fp=0xc44edef338 sp=0xc44edef318
runtime.mapaccess1(0xdfa680, 0xc46637e480, 0xc44edef41c, 0xc4647bac60)
	/home/travis/.gimme/versions/go1.8.linux.amd64/src/runtime/hashmap.go:319 +0x23a fp=0xc44edef380 sp=0xc44edef338
github.com/ethereumproject/go-ethereum/core/state.(*StateDB).GetStateObject(0xc42781a0f0, 0x29688d6b8aa2fa70, 0xa2c96ed249e6b1a4, 0xc413a49ba3, 0xd52710)
	/home/travis/gopath/src/github.com/ethereumproject/go-ethereum/core/state/statedb.go:372 +0xab fp=0xc44edef598 sp=0xc44edef380
github.com/ethereumproject/go-ethereum/core/state.(*StateDB).GetNonce(0xc42781a0f0, 0x29688d6b8aa2fa70, 0xa2c96ed249e6b1a4, 0x13a49ba3, 0xc44edef600)
	/home/travis/gopath/src/github.com/ethereumproject/go-ethereum/core/state/statedb.go:230 +0x3f fp=0xc44edef5d0 sp=0xc44edef598
github.com/ethereumproject/go-ethereum/core/state.(*ManagedState).GetNonce(0xc46637e570, 0x29688d6b8aa2fa70, 0xa2c96ed249e6b1a4, 0xc413a49ba3, 0x0)
	/home/travis/gopath/src/github.com/ethereumproject/go-ethereum/core/state/managed_state.go:94 +0x13f fp=0xc44edef638 sp=0xc44edef5d0
github.com/ethereumproject/go-ethereum/core.(*TxPool).checkQueue(0xc4202f90e0)
	/home/travis/gopath/src/github.com/ethereumproject/go-ethereum/core/tx_pool.go:474 +0x1c1 fp=0xc44edef9d8 sp=0xc44edef638
github.com/ethereumproject/go-ethereum/core.(*TxPool).GetTransactions(0xc4202f90e0, 0x0, 0x0, 0x0)
	/home/travis/gopath/src/github.com/ethereumproject/go-ethereum/core/tx_pool.go:396 +0xcc fp=0xc44edefa88 sp=0xc44edef9d8
github.com/ethereumproject/go-ethereum/eth.(*ProtocolManager).syncTransactions(0xc4202d1ba0, 0xc420634240)
	/home/travis/gopath/src/github.com/ethereumproject/go-ethereum/eth/sync.go:47 +0x49 fp=0xc44edefb70 sp=0xc44edefa88
github.com/ethereumproject/go-ethereum/eth.(*ProtocolManager).handle(0xc4202d1ba0, 0xc420634240, 0x0, 0x0)
	/home/travis/gopath/src/github.com/ethereumproject/go-ethereum/eth/handler.go:278 +0x6c7 fp=0xc44edefdf0 sp=0xc44edefb70
github.com/ethereumproject/go-ethereum/eth.NewProtocolManager.func1(0xc4304c63c0, 0x16d5bc0, 0xc4663abb20, 0x0, 0x0)
	/home/travis/gopath/src/github.com/ethereumproject/go-ethereum/eth/handler.go:134 +0x17f fp=0xc44edefee8 sp=0xc44edefdf0
github.com/ethereumproject/go-ethereum/p2p.(*Peer).startProtocols.func1(0xc4663abb20, 0xc4304c63c0)
	/home/travis/gopath/src/github.com/ethereumproject/go-ethereum/p2p/peer.go:303 +0x5d fp=0xc44edeffd0 sp=0xc44edefee8
runtime.goexit()
	/home/travis/.gimme/versions/go1.8.linux.amd64/src/runtime/asm_amd64.s:2197 +0x1 fp=0xc44edeffd8 sp=0xc44edeffd0
created by github.com/ethereumproject/go-ethereum/p2p.(*Peer).startProtocols
	/home/travis/gopath/src/github.com/ethereumproject/go-ethereum/p2p/peer.go:312 +0x238

goroutine 1 [chan receive, 14 minutes]:
github.com/ethereumproject/go-ethereum/node.(*Node).Wait(0xc4200ac500)
	/home/travis/gopath/src/github.com/ethereumproject/go-ethereum/node/node.go:491 +0x8b
main.geth(0xc420322780, 0x0, 0x0)
	/home/travis/gopath/src/github.com/ethereumproject/go-ethereum/cmd/geth/main.go:257 +0x6e
reflect.Value.call(0xdd2820, 0x1003d90, 0x13, 0xeebd68, 0x4, 0xc42005dd20, 0x1, 0x1, 0xc42005dca8, 0xed3700, ...)
	/home/travis/.gimme/versions/go1.8.linux.amd64/src/reflect/value.go:434 +0x91f
reflect.Value.Call(0xdd2820, 0x1003d90, 0x13, 0xc42005dd20, 0x1, 0x1, 0xc4200601e0, 0xc420022000, 0xeefa2c)
	/home/travis/.gimme/versions/go1.8.linux.amd64/src/reflect/value.go:302 +0xa4
github.com/ethereumproject/go-ethereum/vendor/gopkg.in/urfave/cli%2ev1.HandleAction(0xdd2820, 0x1003d90, 0xc420322780, 0x0, 0x0)
	/home/travis/gopath/src/github.com/ethereumproject/go-ethereum/vendor/gopkg.in/urfave/cli.v1/app.go:480 +0x198
github.com/ethereumproject/go-ethereum/vendor/gopkg.in/urfave/cli%2ev1.(*App).Run(0xc42015d800, 0xc4200100c0, 0xc, 0xc, 0x0, 0x0)
	/home/travis/gopath/src/github.com/ethereumproject/go-ethereum/vendor/gopkg.in/urfave/cli.v1/app.go:241 +0x560
main.main()
	/home/travis/gopath/src/github.com/ethereumproject/go-ethereum/cmd/geth/main.go:245 +0x57

goroutine 17 [syscall, 14 minutes, locked to thread]:

....

### PR DESCRIPTION
**Pull request by whilei**
solution: use mutex lock

fixes #284, again

### CODE REVIEW

File: core/state/statedb.go [hunk 1]
```diff
@@ -349,12 +349,15 @@ func (self *StateDB) deleteStateObject(stateObject *StateObject) {
 // Retrieve a state object given my the address. Returns nil if not found.
 func (self *StateDB) getStateObject(addr common.Address) (stateObject *StateObject) {
 	// Prefer 'live' objects.
+	self.lock.Lock()
 	if obj := self.stateObjects[addr]; obj != nil {
+		self.lock.Unlock()
 		if obj.deleted {
```
Comment (r8d8) [inline core/state/statedb.go]: Сan it be changed here, I mean after `Unlock` and before `return`?

Comment (whilei) [reply core/state/statedb.go]: Sorry.. what do you mean?

Comment (whilei) [reply core/state/statedb.go]: (One could use a `defer self.lock.Unlock()` to unlock before any `return` point, but that would make the map possibly be locked for the rest of the logic after just reading the value) --

Comment (whilei) [reply core/state/statedb.go]: @r8d8 Maybe I see what you mean. Just pushed a small refactor, I think is better.

**Comment by whilei** [comment]
@tzdybal If you have a free second, curious your thoughts on this too. As documented in [the original PR](https://github.com/ethereumproject/go-ethereum/pull/285), I'm still wondering how and why there aren't fewer locks in this file and package. Is it developer laziness? Is it ok and there's just some concept I'm missing? 

Since the original change appears to have fixed the reported issue, I'm OK merging this as-is, but just putting the question/seed-of-doubt on the table for consideration.

**Comment by tzdybal** [comment]
@whilei I don't understand it. It looks like patching ad hoc, just where it actually crashes. Lock in `Copy` is oldest - this is rather long operation, chances for crash are the biggest.

**Comment by whilei** [comment]
Yea.. sketchily ad hoc was my impression too. It seemed to me like the code was somehow usually/luckily "getting away" with not having more locks around the map r/w's. 

I just threw the patch in there where it crashed because I assumed that I didn't understand why my lock was the first around so many map interactions.

**Comment by whilei** [comment]
I'm going to go ahead and merge as-is, then we can maybe come back later and add a more consistent locking strategy. Or at least be aware that we're living dangerously.

**Pull request merged**